command/sync.go

```diff
@@ -0,0 +1,501 @@
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var sourceObjects []*storage.Object
+	var destObjects []*storage.Object
+	var wg sync.WaitGroup
+
+	// get source objects.
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		srcObjectChannel := sourceClient.List(ctx, srcurl, false)
+		for srcObject := range srcObjectChannel {
+			if s.shouldSkipObject(srcObject, true) {
+				continue
+			}
+			sourceObjects = append(sourceObjects, srcObject)
+		}
+	}()
+
+	// add * to end of destination string, to get all objects recursively.
+	var destinationURLPath string
+	if strings.HasSuffix(s.dst, "/") {
+		destinationURLPath = s.dst + "*"
+	} else {
+		destinationURLPath = s.dst + "/*"
+	}
+
+	destObjectsURL, err := url.New(destinationURLPath)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// get destination objects.
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		destObjectsChannel := destClient.List(ctx, destObjectsURL, false)
+		for destObject := range destObjectsChannel {
+			if s.shouldSkipObject(destObject, false) {
+				continue
+			}
+			destObjects = append(destObjects, destObject)
+		}
+	}()
+
+	// wait until source and destination objects are fetched.
+	wg.Wait()
+	return sourceObjects, destObjects, nil
+}
+
+// PlanRun prepares the commands and passes it to reader for run command.
+func (s Sync) PlanRun(ctx context.Context, onlySource, onlyDest []*url.URL, common []*CommonObject,
+	dsturl *url.URL, transferManager transfer.Manager, strategy strategy.Strategy, w *io.PipeWriter, errPlan *error) {
+	// only source objects.
+	for _, srcurl := range onlySource {
+		var curDestURL *url.URL
+		switch {
+		case !srcurl.IsRemote() && dsturl.IsRemote(): // local->remote
+			curDestURL = transferManager.PrepareRemoteDestination(srcurl, dsturl)
+		case srcurl.IsRemote() && !dsturl.IsRemote(): // remote->local
+			url, err := transferManager.PrepareLocalDestination(ctx, srcurl, dsturl)
+			curDestURL = url
+			if err != nil {
+				printError(s.fullCommand, s.op, err)
+				*errPlan = multierror.Append(*errPlan, err)
+				continue
+			}
+		case srcurl.IsRemote() && dsturl.IsRemote(): // remote->remote
+			curDestURL = transferManager.PrepareRemoteDestination(srcurl, dsturl)
+		default:
+			panic("unexpected src-dst pair")
+		}
+		command := fmt.Sprintf("cp %v %v\n", srcurl, curDestURL)
+		fmt.Fprint(w, command)
+	}
+
+	// for common objects
+	for _, commonObject := range common {
+		sourceObject, destObject := commonObject.src, commonObject.dst
+		curSourceURL, curDestURL := sourceObject.URL, destObject.URL
+		err := strategy.Compare(sourceObject, destObject) // check if object should be copied.
+		if err != nil {
+			if errorpkg.IsWarning(err) {
+				printDebug(s.op, curSourceURL, curDestURL, err)
+				continue
+			}
+		}
+
+		command := fmt.Sprintf("cp %v %v\n", curSourceURL, curDestURL)
+		fmt.Fprint(w, command)
+	}
+
+	// for only destination objects.
+	if s.delete { // if delete is set
+		urlpaths := GenerateRemovePath(onlyDest)
```

Comment:
What happens if the `onlyDest` is empty? `rm %v` would be `rm ` and it is gonna fail.

Reply:
add an if condition, if there are items to delete, then it will call `rm`.

Reply:
Could you write a test for this case please?

Reply:
I'm adding the test myself Hasan.